### PR TITLE
Nnvm RNN op with icc and mkldnn memory cache 

### PR DESCRIPTION
**Pull request by lihaofd**
## Description ##
(Brief description on what this PR is about)

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [ ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ ] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [ ] Feature1, tests, (and when applicable, API doc)
- [ ] Feature2, tests, (and when applicable, API doc)

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 1]
```diff
@@ -0,0 +1,589 @@
...
+#include <vector>
+#include "../../rnn_impl.h"
+#include "mkldnn.hpp"
+#include "./mkldnn_base-inl.h"
+
+namespace mxnet {
+namespace op {
+
+namespace mkldnn_rnn_enum {
+  enum RNNModeType {kRnnRelu, kRnnTanh, kLstm, kGru};
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: it's already defined in operator/rnn-inl.h.

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 2]
```diff
@@ -0,0 +1,589 @@
...
+}
+
+template <typename DType>
+mkldnn::memory::data_type GetMKLDNNDataType() {
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: Looks strange and unsafe for me. Please refer the example in https://en.cppreference.com/w/cpp/language/typeid .

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 3]
```diff
@@ -0,0 +1,589 @@
...
+      algo = algorithm::vanilla_lstm;
+      break;
+    default:
+      LOG(FATAL) << "unsupported RNN mode" << mode;
+      break;
+  }
+  return algo;
+}
+
+void ReorderForWeight(mkldnn::memory src,
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: Why `ReorderForWeight`? Looks like a general reorder function, not only for weight.

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 4]
```diff
@@ -0,0 +1,589 @@
...
+                      mkldnn::memory dst) {
+  auto r = mkldnn::reorder(src, dst);
+  stream(stream::kind::eager).submit({r}).wait();
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: what's stream? If it's `mkldnn::stream`, you need call `MKLDNNStream::Get()`.

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 5]
```diff
@@ -0,0 +1,589 @@
...
+    srcs.push_back(src_memory);
+  }
+  std::vector<primitive::at> inputs;
+  for (size_t i = 0; i < srcs_cds.size(); i++) {
+    inputs.push_back(srcs[i]);
+  }
+
+  auto dst_desc = mkldnn::memory::desc(dst_cds, mkldnn_dtype, dst_format);
+  auto concat_pd = mkldnn::concat::primitive_desc(dst_desc, concat_dimension, srcs_pd);
+  auto dst = mkldnn::memory(concat_pd.dst_primitive_desc());
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: you will allocate temporal memory here...

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 6]
```diff
@@ -0,0 +1,589 @@
...
+}
+
+template <typename DType>
+mkldnn::memory ConcatData(mkldnn::memory::format src_format,
+                          mkldnn::memory::format dst_format,
+                          std::vector<mkldnn::memory::dims> srcs_cds,
+                          mkldnn::memory::dims dst_cds,
+                          mkldnn::memory::data_type mkldnn_dtype,
+                          int concat_dimension,
+                          std::vector<DType*> srcs_data) {
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: It's not necessary to have `DType` here. Use 'void*' then you needn't have this function as a template function.

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 7]
```diff
@@ -0,0 +1,589 @@
...
+  auto cpu_engine = CpuEngine::Get()->get_engine();
+  std::vector<mkldnn::memory::primitive_desc> srcs_pd;
+  std::vector<mkldnn::memory> srcs;
+  for (size_t i = 0; i < srcs_cds.size(); i++) {
+    auto desc = memory::desc(srcs_cds[i], mkldnn_dtype, src_format);
+    auto mpd = memory::primitive_desc(desc, cpu_engine);
+    auto src_memory = mkldnn::memory({desc, cpu_engine}, srcs_data[i]);
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: atuo src_memory = mkldnn::memory(mpd, srcs_data[i]);

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 8]
```diff
@@ -0,0 +1,589 @@
...
+}
+
+template <typename DType>
+void MKLDNNRNNForward(bool state_outputs,
+                      int L,
+                      int D,
+                      const int T,
+                      const int N,
+                      int I,
+                      const int H,
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: Why T, N, H are const but L, D, I are not const? Any special consideration?

**Review by TaoLv — COMMENTED** [review]
Reviewed part of the change. Will review the rest once get time. Ping @ciyongch.

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 1 above)
Comment (lihaofd) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: fixed

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 2 above)
Comment (lihaofd) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: fixed

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 3 above)
Comment (lihaofd) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: fixed

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 4 above)
Comment (lihaofd) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: fixed

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 6 above)
Comment (lihaofd) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: fixed

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 7 above)
Comment (lihaofd) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: fixed

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 8 above)
Comment (lihaofd) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: fixed

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 2 above)
Comment (ciyongch) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: There's already a function "get_mkldnn_type()" in mkldnn_base-inl.h, can we just reuse that function instead of creating a similar one?

File: src/operator/rnn-inl.h [hunk 9]
```diff
@@ -39,6 +39,7 @@
 #include "./math_functions-inl.h"
 #include "./operator_common.h"
 #include "./rnn_impl.h"
+#include "./nn/mkldnn/mkldnn_rnn_impl.h"
```
Comment (ciyongch) [inline src/operator/rnn-inl.h]: add "#if MXNET_USE_MKLDNN ==1" for mkldnn related header file.

File: src/operator/rnn-inl.h [hunk 10]
```diff
@@ -620,203 +638,269 @@ class RNNOp : public Operator{
...
+    it = ins_ret.first;
   }
+  return *it->second;
+}
 
-  void Init(const std::vector<std::pair<std::string, std::string> >& kwargs) override {
-    param_.Init(kwargs);
+#if MXNET_USE_MKLDNN == 1
+template<typename DType>
+static RNNOp<DType> &GetMKLDNNRNNOp(const RNNParam &param,
```
Comment (ciyongch) [inline src/operator/rnn-inl.h]: move MKLDNN functions to mkldnn related files.

File: src/operator/rnn-inl.h [hunk 11]
```diff
@@ -620,203 +638,269 @@ class RNNOp : public Operator{
...
+  RNNSignature key(param);
+  key.AddSign(weight);
+  key.AddSign(ctx.dev_id);
+
+  auto it = ops.find(key);
+  if (it == ops.end()) {
+    std::shared_ptr<RNNOp<DType>> op(new RNNOp<DType>(param));
+    auto ins_ret = ops.insert(std::pair<RNNSignature, std::shared_ptr<RNNOp<DType> > >(key, op));
+    CHECK(ins_ret.second);
+    it = ins_ret.first;
```
Comment (ciyongch) [inline src/operator/rnn-inl.h]: refer mkldnn_convolution.cc to use AddToCache() function here.
Why not add data/output here?

File: src/operator/rnn.cc [hunk 12]
```diff
@@ -21,89 +21,200 @@
...
-            \end{array}
+static bool RNNShape(const nnvm::NodeAttrs& attrs,
+                     std::vector<TShape> *in_shape,
+                     std::vector<TShape> *out_shape) {
+  const RNNParam& param_ = nnvm::get<RNNParam>(attrs.parsed);
+  using namespace mshadow;
+  if (param_.mode == rnn_enum::kLstm) {
+    CHECK_EQ(in_shape->size(), 4U) << "Input:[data, parameters, state, cell_state]";
+  } else {
+    CHECK_EQ(in_shape->size(), 3U) << "Input:[data, parameters, state]";
```
Comment (ciyongch) [inline src/operator/rnn.cc]: Better to make the error info more comprehensible, such as expected 4 inputs ..... but got x inputs?

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 13]
```diff
@@ -0,0 +1,589 @@
...
+      datatype = mkldnn::memory::data_type::u8;
+      break;
+    default:
+      datatype = mkldnn::memory::data_type::data_undef;
+      break;
+  }
+  return datatype;
+}
+
+mkldnn::algorithm GetMKLDNNAlgo(int mode,
```
Comment (ciyongch) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: Suggest to use `GetMKLDNNRNNAlgo` instead of current one. Also make `mode` to a const param.

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 14]
```diff
@@ -0,0 +1,589 @@
...
+    src_wh0 = mkldnn::memory(mpd_h, wh_0);
+  } else {
+    std::vector<DType*> srcs_data1;
+    srcs_data1.push_back(wh_0);
+    srcs_data1.push_back(back_wh_0);
+    src_wh0 =
+        ConcatData(mkldnn::memory::format::ldgoi, mkldnn::memory::format::ldgoi,
+            {weights_iter_r_tz_0, weights_iter_r_tz_0}, weights_iter_tz_0,
+             mkldnn_dtype, 1, srcs_data1);
+  }
```
Comment (ciyongch) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: combine these two if closure into one.

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 15]
```diff
@@ -0,0 +1,589 @@
...
+  auto user_bias_md_0 = mkldnn::memory::desc({ bias_tz_0 },
+      mkldnn_dtype, mkldnn::memory::format::ldgo);
+  auto dst_layer_md_0 = mkldnn::memory::desc(
+      { dst_layer_tz_0 }, mkldnn_dtype, mkldnn::memory::format::tnc);
+  auto dst_iter_md_0 = mkldnn::memory::desc(
+      { dst_iter_tz_0 }, mkldnn_dtype, mkldnn::memory::format::ldsnc);
+
+  auto user_src_layer_memory_0 = mkldnn::memory(
+      { user_src_layer_md_0, cpu_engine }, x_0);
+
```
Comment (ciyongch) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: can we use the way  `SetNewMem` as current MKLDNN ops?

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 16]
```diff
@@ -0,0 +1,589 @@
...
+        offset2 = (nstates + 1) * single_cell_size;
+        #pragma omp parallel for num_threads(omp_threads)
+        for (int n = 0; n < single_cell_size; n++) {
+          cy_ptr[n + single_cell_size] = dst_hcy_0[n + offset2];
+        }
+      }
+    }
+  }
+  //  go to next L - 1 layers.
+  //  If D = 2, do it layer by layer. If D = 1, fused L - 1 layers
```
Comment (ciyongch) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: For the case of `I != H`, I think the two stage is reasonable due to MKLDNN API restriction.
For the case of `I == H`, do you think it's able to combine the two stage into one? If so, only one primitive is needed.

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 17]
```diff
@@ -0,0 +1,589 @@
...
+  memory::dims src_layer_tz_0 = {T, N, I};
+  memory::dims weights_layer_tz_0 = {1, D, I, ngates, H};  //  ldigo
+  memory::dims weights_layer_r_tz_0 = {1, 1, I, ngates, H};  //  ldigo for reorder
+  memory::dims weights_iter_tz_0 = {1, D, H, ngates, H};  //  ldigo
+  memory::dims weights_iter_r_tz_0 = {1, 1, H, ngates, H};  //  ldigo for reorder
+  memory::dims bias_tz_0 = {1, D, ngates, H};
+  memory::dims dst_layer_tz_0 = {T, N, D * H};
+  memory::dims src_iter_tz_0 = {1, D, nstates, N, H};  //  ldsnc
+  memory::dims dst_iter_tz_0 = {1, D, nstates, N, H};  //  ldsnc
+  int offset1 = 0, offset2 = 0;
```
Comment (ciyongch) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h:140]: minor comments for the variables naming, please use more meaningful names. maybe change `back_xxx` to `l2r_xxx` or `r2l_xxx`?

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 18]
```diff
@@ -0,0 +1,589 @@
...
+  DType* back_w_ptr = w_ptr;
+  DType* back_b_ptr = b_ptr;
+  DType* wx_0 = w_ptr;  //  ngates * H, I
+  DType* wh_0 = w_ptr + I * H * ngates;  //  ngates * H, H
+  if (D == 2) {
+    back_w_ptr = w_ptr + ngates * H * (I + H);
+    back_b_ptr = b_ptr + single_b_size * 2;
+  }
+  DType* back_wx_0 = back_w_ptr;
+  DType* back_wh_0 = back_w_ptr + I * H * ngates;
```
Comment (ciyongch) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: move all the assignment of back_xxx to `if (D == 2)` closure, and with default value here is enough.

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 5 above)
Comment (lihaofd) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: fixed

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 18 above)
Comment (lihaofd) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: will trying it, but some of them have ordered definition and can't move them all

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 17 above)
Comment (lihaofd) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h:140]: No, in non_mkldnn path, the naming of GRU/LSTM/vRNN are based on back_xxx, it is better to keep them same

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 14 above)
Comment (lihaofd) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: fixed

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h (hunk 2 above)
Comment (lihaofd) [reply src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: fixed

File: src/operator/rnn-inl.h (hunk 9 above)
Comment (lihaofd) [reply src/operator/rnn-inl.h]: fixed

**Comment by TaoLv** [comment]
Cannot see your fix. Please push them here. @lihaofd

**Comment by lihaofd** [comment]
@TaoLv I am still fixing the code, if I pushed the code now, many of comment's relate code can't be found easily, I will push the code fix after I fix them all.

File: src/operator/rnn-inl.h (hunk 10 above)
Comment (lihaofd) [reply src/operator/rnn-inl.h]: many of class and struct are shared here, if moving all to mkldnn related file, it will create many repeated codes

**Comment by TaoLv** [comment]
I see. So it's "will fix" rather than "fixed".

File: src/operator/rnn.cc [hunk 19]
```diff
@@ -21,89 +21,200 @@
...
+                    std::vector<int> *out_type) {
+  const RNNParam& param_ = nnvm::get<RNNParam>(attrs.parsed);
+  CHECK_GE(in_type->size(), 1U);
+  int dtype = (*in_type)[0];
+  CHECK_NE(dtype, -1) << "First input must have specified type";
+  for (size_t i = 0; i < in_type->size(); ++i) {
+    if ((*in_type)[i] == -1) {
+      (*in_type)[i] = dtype;
+    } else {
+      UNIFORM_TYPE_CHECK((*in_type)[i], dtype, ListArguments(param_)[i]);
```
Comment (TaoLv) [inline src/operator/rnn.cc:111]: May exceed the boundary of list if i > 3.

File: src/operator/rnn.cc [hunk 20]
```diff
@@ -21,89 +21,200 @@
...
-Gated Recurrent Unit - Cho et al. 2014. http://arxiv.org/abs/1406.1078
+inline static bool RNNStorageType(const nnvm::NodeAttrs& attrs,
+                                  const int dev_mask,
+                                  DispatchMode* dispatch_mode,
+                                  std::vector<int> *in_attrs,
+                                  std::vector<int> *out_attrs) {
+  DispatchMode wanted_mode = DispatchMode::kFCompute;
+  #if MXNET_USE_MKLDNN == 1
+    wanted_mode = DispatchMode::kFComputeEx;
+  #endif
```
Comment (TaoLv) [inline src/operator/rnn.cc:135]: it doesn't work for gpu.

File: src/operator/rnn-inl.h (hunk 10 above)
Comment (TaoLv) [reply src/operator/rnn-inl.h]: Agree to move MKLDNN code to mkldnn_rnn_impl.h. Please refer what we did for other operators.

File: src/operator/rnn-inl.h [hunk 21]
```diff
@@ -620,203 +638,269 @@ class RNNOp : public Operator{
...
-    TShape oshape = dshape;
-    if (param_.projection_size.has_value()) {
-      oshape[2] = numDirections * param_.projection_size.value();
+#if MXNET_USE_MKLDNN == 1
+void RNNComputeExCPU(const nnvm::NodeAttrs& attrs,
+                     const OpContext &ctx,
+                     const std::vector<NDArray> &inputs,
+                     const std::vector<OpReqType> &req,
+                     const std::vector<NDArray> &outputs) {
+  RNNParam& param = (RNNParam&)nnvm::get<RNNParam>(attrs.parsed);
```
Comment (TaoLv) [inline src/operator/rnn-inl.h]: Many code in this function should be moved to MKLDNNRNNForward.

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 22]
```diff
@@ -0,0 +1,589 @@
...
+
+  rnn_cell::desc cell_0(nalgorithm);
+  rnn_forward::desc layer_desc_0(prop_kind::forward_inference, cell_0,
+      D == 1 ? rnn_direction::unidirectional : rnn_direction::bidirectional_concat,
+      user_src_layer_md_0, user_src_iter_md_0, user_weight_layer_md_0, user_weight_iter_md_0,
+      user_bias_md_0, dst_layer_md_0, dst_iter_md_0);
+
+  auto prim_desc_0 = mkldnn::rnn_forward::primitive_desc(layer_desc_0, cpu_engine);
+  auto dst_layer_memory_0 = mkldnn::memory(prim_desc_0.dst_layer_primitive_desc());
+  auto dst_iter_memory_0 = mkldnn::memory(prim_desc_0.dst_iter_primitive_desc());
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: It will allocate memory.

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 23]
```diff
@@ -0,0 +1,589 @@
...
+  dst_layer_memory_0.set_data_handle(y);
+
+  rnn_net_0.push_back(
+      rnn_forward(prim_desc_0, user_src_layer_memory_0,
+                  user_src_iter_memory_0, (*weight_bias_memory)[0],
+                  (*weight_bias_memory)[1], (*weight_bias_memory)[2],
+                  dst_layer_memory_0, dst_iter_memory_0, null_memory_));
+  stream(stream::kind::eager).submit(rnn_net_0).wait();
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: MKLDNNStream

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 24]
```diff
@@ -0,0 +1,589 @@
...
+        { dst_iter_tz }, mkldnn_dtype, mkldnn::memory::format::ldsnc);
+    rnn_cell::desc cell(nalgorithm);
+    rnn_forward::desc layer_desc(prop_kind::forward_inference, cell,
+        rnn_direction::bidirectional_concat,
+        user_src_layer_md, user_src_iter_md, user_weight_layer_md, user_weight_iter_md,
+        user_bias_md, dst_layer_md, dst_iter_md);
+    auto prim_desc = mkldnn::rnn_forward::primitive_desc(layer_desc, cpu_engine);
+    auto dst_layer_memory = mkldnn::memory(prim_desc.dst_layer_primitive_desc());
+    dst_layer_memory.set_data_handle(y);
+    auto dst_iter_memory = mkldnn::memory(prim_desc.dst_iter_primitive_desc());
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: it will allocate memory.

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 25]
```diff
@@ -0,0 +1,589 @@
...
+        { dst_iter_tz }, mkldnn_dtype, mkldnn::memory::format::ldsnc);
+    rnn_cell::desc cell(nalgorithm);
+    rnn_forward::desc layer_desc(prop_kind::forward_inference, cell,
+        rnn_direction::unidirectional,
+        user_src_layer_md, user_src_iter_md, user_weight_layer_md, user_weight_iter_md,
+        user_bias_md, dst_layer_md, dst_iter_md);
+    auto prim_desc = mkldnn::rnn_forward::primitive_desc(layer_desc, cpu_engine);
+    auto dst_layer_memory = mkldnn::memory(prim_desc.dst_layer_primitive_desc());
+    dst_layer_memory.set_data_handle(y);
+    auto dst_iter_memory = mkldnn::memory(prim_desc.dst_iter_primitive_desc());
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: memory allocation again.

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 26]
```diff
@@ -0,0 +1,589 @@
...
+    for (int j = 0; j < (L - 1) * single_b_size; j++) {
+      int k = j / single_b_size;
+      user_bias_f[j] = b_ptr[j + k * single_b_size] + b_ptr[j + k * single_b_size + single_b_size];
+    }
+    rnn_net2.push_back(
+        rnn_forward(prim_desc, user_src_layer_memory_l,
+                    user_src_iter_memory, (*weight_bias_memory)[3],
+                    (*weight_bias_memory)[4], (*weight_bias_memory)[5],
+                    dst_layer_memory, dst_iter_memory, null_memory_));
+    stream(stream::kind::eager).submit(rnn_net2).wait();
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: MKLDNNStream

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 27]
```diff
@@ -0,0 +1,589 @@
...
+  auto concat_pd = mkldnn::concat::primitive_desc(dst_desc, concat_dimension, srcs_pd);
+  auto dst = mkldnn::memory(concat_pd.dst_primitive_desc());
+  
+  auto c = mkldnn::concat(concat_pd, inputs, dst);
+  stream(stream::kind::eager).submit({c}).wait();
+  return dst;
+}
+
+template <typename DType>
+void MKLDNNRNNForward(bool state_outputs,
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: This function is too long to review and maintain. Can you split it into several functions and combine them together here. I guess some blocks can also be shared when we implement backward.

File: src/operator/nn/mkldnn/mkldnn_rnn_impl.h [hunk 28]
```diff
@@ -0,0 +1,589 @@
...
+  for (size_t i = 0; i < srcs_cds.size(); i++) {
+    inputs.push_back(srcs[i]);
+  }
+
+  auto dst_desc = mkldnn::memory::desc(dst_cds, mkldnn_dtype, dst_format);
+  auto concat_pd = mkldnn::concat::primitive_desc(dst_desc, concat_dimension, srcs_pd);
+  auto dst = mkldnn::memory(concat_pd.dst_primitive_desc());
+  
+  auto c = mkldnn::concat(concat_pd, inputs, dst);
+  stream(stream::kind::eager).submit({c}).wait();
```
Comment (TaoLv) [inline src/operator/nn/mkldnn/mkldnn_rnn_impl.h]: MKLDNNStream

**Comment by TaoLv** [comment]
No need to wait for all comments addressed. Github will help to maintain the review history.
